docs: review for OCM plugin (#291)
janus-idp/backstage-plugins: plugins/ocm/README.md

# Open Cluster Management plugin for Backstage

> The Open Cluster Management (OCM) plugin integrates your Backstage instance with the `MultiClusterHub` and `MultiCluster` engines of Open Cluster Management.

Reviewer: ```suggestion
The Open Cluster Management (OCM) plugin integrates your Backstage instance with the `MultiClusterHub` and `MultiCluster` engines of OCM.
```

can be OCM in the second instance since it's defined earlier in the sentence?

Author: Done

## Prerequisites

- OCM is deployed and configured on a Kubernetes cluster.
- [Kubernetes plugin for Backstage](https://backstage.io/docs/features/kubernetes/overview) is installed.

> - Kubernetes plugin is configured and able to connect to the Hub cluster for OCM using a `ServiceAccount`. The cluster is accessed using Backstage backend, therefore, the cluster requires a permanent connection with constant RBAC scope.

Reviewer: Hub is capitalized here, but not earlier in the docs. Should it be "hub" or "Hub" throughout? It's good to standardize on capitalization. Also, earlier in the docs "service account" is used, and here it's `ServiceAccount`. I would stick with the general usage "service account" unless you're referring to a specific `ServiceAccount` object.

Author: Changed Hub to hub. And We are referring to ServiceAccount object, so I'd keep as-is.

> - The `ClusterRole` must be granted to `ServiceAccount` accessing the Hub cluster as follows:

Reviewer: ```suggestion
- A `ClusterRole` must be created for the service account, so that the service account can access the Hub cluster:
```

Author: The suggestion actually changes the meaning of the sentence:

A ClusterRole (is already created) is not created but granted to ServiceAccount that accesses the hub cluster. This action does not enable the service account to access the hub cluster.

  ```yaml
  kind: ClusterRole
  apiVersion: rbac.authorization.k8s.io/v1
  metadata:
    name: backstage-ocm-plugin
  rules:
    - apiGroups:
        - cluster.open-cluster-management.io
      resources:
        - managedclusters
      verbs:
        - get
        - watch
        - list
    - apiGroups:
        - internal.open-cluster-management.io
      resources:
        - managedclusterinfos
      verbs:
        - get
        - watch
        - list
  ```

> ## Capabilities

Reviewer: I would move information about capabilities to before the prerequisites (abstract / intro)

Author: Good point. Thank you. 
Moved it before Prerequisites.

The OCM plugin has the following capabilities:

- All clusters represented as `ManagedCluster` in `MultiClusterHub` or MCE are discovered and imported into the Backstage catalog, such as:
  - Entity is defined as `kind: Resource` with `spec.type` set to `kubernetes-cluster`.

>   - Links to the OCP console, OCM console, and OpenShift Cluster Manager are provided in `metadata.links`.

Reviewer: ```suggestion
  - Links to the OpenShift Container Platform console, OCM console, and OCM are provided in `metadata.links`.
```

Reviewer: Check on official usage of OpenShift naming maybe? Not sure how we want to refer to this in upstream docs like this

> - Shows real-time data from OCM on the Resource entity page including:

Reviewer: ```suggestion
- Shows real-time data from OCM on the Resource entity page, including:
```

  - Cluster current status (up or down)

>   - Cluster details (console link, OCP, and Kubernetes version)

Reviewer: same comment re OCP

Author: Defined OCP as OpenShift Container Platform (OCP) in the previous sentence.

> - Kubernetes plugin is configured and can connect to the hub cluster for OCM using a `ServiceAccount`. The cluster is accessed using Backstage backend, therefore, the cluster requires a permanent connection with constant RBAC scope.

Reviewer 2: I think we should clarify this. This is obsolete information. Kubernetes plugin for Backstage is not a prerequisite. It's optional. OCM plugin can either reference a cluster directly or defer to the Kubernetes plugin for details. We manage the connection ourselves.

Let's remove it from prerequisites and instead add a note below (in the config section) describing this optional behavior.

Author: Removed it and added in the config section.

---

### Set up `@janus-idp/backstage-plugin-ocm-backend` package

1. Install the OCM backend plugin using the following command:

   ```sh
   yarn workspace backend add @janus-idp/backstage-plugin-ocm-backend
   ```

2. Configure the OCM backend plugin using one of the following configurations:

   - The OCM configuration provides the information about your hub. To use the OCM configuration, add the following code to your `app-config.yaml` file:

     ```yaml
     # app-config.yaml
     catalog:
       providers:
         ocm:
           env: # Key is reflected as provider ID. Defines and claims plugin instance ownership of entities
             name: # Name that the hub cluster will assume in Backstage Catalog (in OCM this is always local-cluster which can be confusing)
             url: # Url of the hub cluster API endpoint
             serviceAccountToken: # Token used for querying data from the hub
             skipTLSVerify: # Skip TLS certificate verification, defaults to false
             caData: # Base64-encoded CA bundle in PEM format
     ```

>    - You can bind the hub configuration to the Kubernetes configuration by providing the name of the hub as follows:

Reviewer 2: ```suggestion
   - If you also have Backstage Kubernetes plugin installed and configured to connect to the Hub cluster, you can defer to this configuration by providing the name of the hub as follows. Please make sure Backstage uses a `ServiceAccount` token and was granted permissions described above.
```

See the previous review comment. Feel free to reword it. 🙂

Author: Done

     ```yaml
     # app-config.yaml
     kubernetes:
       serviceLocatorMethod:
         type: "multiTenant"
       clusterLocatorMethods:
         - type: "config"
           clusters:
             - name: <cluster-name>
             ...

     catalog:
       providers:
         ocm:
           env: # Key is reflected as provider ID. Defines and claims plugin instance ownership of entities
             kubernetesPluginRef: <cluster-name> # Match the cluster name in kubernetes plugin config
     ```

>      This is useful when you already use a Kubernetes plugin in your Backstage instance. Also, the hub cluster must be connected using `ServiceAccount`.

Reviewer: Same comment re capitalization of hub vs Hub. Same comment as previous re service account vs `ServiceAccount`.

Author: Referring to `ServiceAccount` object, which is similar to our other docs.

Reviewer: in that case I'd specifically say "a `ServiceAccount` object" per the guidance in https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-formatting

Author: Sure, makes sense

     }
   ```

> 5. Import the cluster `Resource` entity provider into `catalog` plugin in `packages/backend/src/plugins/catalog.ts`file:

Reviewer: ```suggestion
5. Import the cluster `Resource` entity provider into the `catalog` plugin in the `packages/backend/src/plugins/catalog.ts`file:
```

         ...
       </ClusterContextProvider>
     ```

   - `ClusterInfoCard`: This is an entity component displaying details of a cluster in a table:

     ```diff
       <ClusterContextProvider>
         ...
     +   <ClusterInfoCard />
         ...
       </ClusterContextProvider>
     ```

   - `ClusterAvailableResourceCard`: This is an entity component displaying the available resources on a cluster. For example, see [`.status.capacity`](https://open-cluster-management.io/concepts/managedcluster/#cluster-heartbeats-and-status) of the `ManagedCluster` resource.

     ```diff
       <ClusterContextProvider>
         ...
     +   <ClusterAvailableResourceCard />
         ...
       </ClusterContextProvider>
     ```

   - `ClusterAllocatableResourceCard`: This is an entity component displaying allocatable resources on a cluster. For example, see [`.status.allocatable`](https://open-cluster-management.io/concepts/managedcluster/#cluster-heartbeats-and-status) of the `ManagedCluster` resource.

     ```diff
       <ClusterContextProvider>
         ...
     +   <ClusterAllocatableResourceCard />
         ...
       </ClusterContextProvider>
     ```

## Development

> If you have installed the OCM plugin to the example application in the repository, then run `yarn start` command to access the plugin in the root directory and then navigate to [/ocm](http://localhost:3000/ocm).

Reviewer: ```suggestion
If you have installed the OCM plugin to the example application in the repository, run the `yarn start` command to access the plugin in the root directory and then navigate to [/ocm](http://localhost:3000/ocm).
```

> To start a development setup in isolation with a faster setup and hot reloads, follow the following steps:

Reviewer: ```suggestion
To start a development setup in isolation with a faster setup and hot reloads, complete the following steps:
```

Author: Done

> 1. Run the `ocm-backend` plugin in the `plugins/ocm-backend` directory by executing the following command:

Reviewer: use consistent language - run vs execute vs enter vs use when referring to commands. I would go with "run the following command" throughout
